sdks/python/apache_beam/runners/interactive/display/pcoll_visualization_test.py

```diff
@@ -0,0 +1,133 @@
+
+# Work around nose tests using Python2 without unittest.mock module.
+try:
+  from unittest.mock import patch
+except ImportError:
+  from mock import patch
+
+
+class PCollVisualizationTest(unittest.TestCase):
+
+  def setUp(self):
+    self._p = beam.Pipeline()
+    # pylint: disable=range-builtin-not-iterating
+    self._pcoll = self._p | 'Create' >> beam.Create(range(1000))
+
+  @unittest.skipIf(sys.version_info < (3, 5, 3),
+                   'PCollVisualization is not supported on Python 2.')
+  def test_raise_error_for_non_pcoll_input(self):
+    class Foo(object):
+      pass
+
+    with self.assertRaises(ValueError) as ctx:
+      pv.PCollVisualization(Foo())
+      self.assertTrue('pcoll should be apache_beam.pvalue.PCollection' in
+                      ctx.exception)
+
+  @unittest.skipIf(sys.version_info < (3, 5, 3),
+                   'PCollVisualization is not supported on Python 2.')
+  def test_pcoll_visualization_generate_unique_display_id(self):
+    pv_1 = pv.PCollVisualization(self._pcoll)
+    pv_2 = pv.PCollVisualization(self._pcoll)
+    self.assertNotEqual(pv_1._dive_display_id, pv_2._dive_display_id)
+    self.assertNotEqual(pv_1._overview_display_id, pv_2._overview_display_id)
+    self.assertNotEqual(pv_1._df_display_id, pv_2._df_display_id)
+
+  @unittest.skipIf(sys.version_info < (3, 5, 3),
+                   'PCollVisualization is not supported on Python 2.')
+  @patch('apache_beam.runners.interactive.display.pcoll_visualization'
+         '.PCollVisualization._to_element_list', lambda x: [1, 2, 3])
+  def test_one_shot_visualization_not_return_handle(self):
+    self.assertIsNone(pv.visualize(self._pcoll))
+
+  def _mock_to_element_list(self):
+    yield [1, 2, 3]
+    yield [1, 2, 3, 4]
+
+  @unittest.skipIf(sys.version_info < (3, 5, 3),
+                   'PCollVisualization is not supported on Python 2.')
+  @patch('apache_beam.runners.interactive.display.pcoll_visualization'
+         '.PCollVisualization._to_element_list', _mock_to_element_list)
+  def test_dynamical_plotting_return_handle(self):
+    h = pv.visualize(self._pcoll, dynamical_plotting_interval=1)
+    self.assertIsInstance(h, timeloop.Timeloop)
+    h.stop()
+
+  @unittest.skipIf(sys.version_info < (3, 5, 3),
+                   'PCollVisualization is not supported on Python 2.')
+  @patch('apache_beam.runners.interactive.display.pcoll_visualization'
+         '.PCollVisualization._to_element_list', _mock_to_element_list)
+  @patch('apache_beam.runners.interactive.display.pcoll_visualization'
+         '.PCollVisualization.display_facets')
+  def test_dynamical_plotting_update_same_display(self,
+                                                  mocked_display_facets):
+    # Starts async dynamical plotting.
+    h = pv.visualize(self._pcoll, dynamical_plotting_interval=0.001)
+    # Blocking so the above async task can execute a few iterations.
+    time.sleep(0.1)
```

Comment:
Please don't use time.sleep() in unit tests. Not only does it slow down the testing, but it also generally leads to flaky tests, and is a bad code smell.

Reply:
Using while loop querying for an ending condition to block then. Also added a timeout to ensure the while loop to never exceed 0.1 second. Even if the logic runs into issue, it will just fail and will not block the test forever.